Add a4x2-package/a4x2-deploy xtasks

faithanalog opened the pull request:
## overview

This PR adds two new xtasks:

- a4x2-package
- a4x2-deploy

### a4x2-package

```
> cargo xtask a4x2-package -h
Generate a tarball with omicron packaged for deployment onto a4x2

Usage: cargo xtask a4x2-package [OPTIONS]

Options:
      --live-tests        Bundle omicron live tests into the output tarball
      --end-to-end-tests  Bundle omicron end-to-end-tests package into the output tarball
  -h, --help              Print help
```

This command will:

- download & build the latest a4x2 from oxidecomputer/testbed
- clone the local working tree to a temporary path & build + package it with a4x2
- if requested, build live-tests
- if requested, build end-to-end tests (but there is no way to run these yet)
- package these all into a tar bundle, at `out/a4x2-package-out.tgz`


### a4x2-deploy

```
> cargo xtask a4x2-deploy -h
Run a4x2 and deploy omicron onto it, and optionally run live-tests and end to end tests

Usage: cargo xtask a4x2-deploy <COMMAND>

Commands:
  start      Start a4x2, deploy a control plane to it, and then leave it running
  stop       Stop a4x2 that was previously launched with the `start` subcommand
  run-tests  Start a4x2, run tests, and then stop a4x2 unless you request otherwise
  help       Print this message or the help of the given subcommand(s)

Options:
  -h, --help  Print help
```

This command will:

- tear down any existing a4x2 created by a previous invocation of a4x2-deploy
- unpack a bundle generated by a4x2-package
- download latest OVMF code & propolis server
- launch a4x2
- ping the control plane until it becomes alive (or a timeout is reached, in which case it considers the launch failed)
- if requested, transfer live tests to the system and execute them
- print information on how to access the server

By default, a4x2-deploy will use a bundle from `out/a4x2-package.tgz`, but an alternative can be specified with `--package <path>`, so you can build a package and send it to someone else for them to run on their machine.


## Known shortcomings

- a4x2 is still a bit flaky. Sometimes I will reboot my machine for good luck.
- a4x2-deploy could and should automatically collect logs from the system when launches fail, but does not yet.
- no support for incremental builds of omicron. a4x2 makes some changes to tomls in the current working directory, which is why we clone the source directory to another path prior to running the a4x2 package step.

Additionally, because of cloning the source to a temporary working path, any changes you have not `git add`ed will not be seen by `a4x2-package`. I consider this either mildly useful or mildly confusing depending on what you're doing in the moment. The jj crew can of course just run `jj` before using `a4x2-package`.

We could opt instead of doing a `cp -a` of the working directory instead of a git clone, but this could be problematic due to copying potentially massive `target/` directories. We could use `rsync` with excludes, but I'm not sure we can rely on it always being present. The `git clone` seemed like the most reliable to me, and it has precedent in the `releng` xtask. But I do find this to be somewhat annoying when iterating on changes.


## Aside on xshell

This PR adds a dependency on [xshell](https://docs.rs/xshell/0.2.7/xshell/index.html), which I use heavily for running system commands. I figured I would take the opportunity to talk about my experience with it, and provide some context that may be helpful when reviewing this PR. I've really enjoyed using it and think it makes it quite pleasant to use rust as a bash replacement.

A few things to keep in mind when reading it, which you can learn from the docs, but I will also highlight here:


### All commands are echoed by default

Similar to `set -x` in bash, commands are echoed as they are run. This is great for logs. Additionally, non-zero exit codes become `Err`s, and command stdout/stderr are sent to the terminal by default, unless you make a point of capturing them.

Command invocations are not echoed when you use `.read()` to capture stdout


### `cmd!()` is *not* sending your command into a shell

`cmd!()` is building a normal rust `Command`! If you see something like

```rust
cmd!(sh, "some command here {with_args}")
```
You may ask "will that escape `with_args`"? The answer is, there is no shell invocation, so there is nothing to escape! It is similar to as if you had done

```rust
Command::new("some").arg("command").arg("here").arg(&with_args)
```

It is mostly convenience around `Command`, with a few other file access shorthands thrown in.


### `sh.pushdir()`

You will see a lot of a pattern

```rust
let _popdir = sh.pushdir(some_dir);
```

`sh` has its own internal working directory it stores (separate from the process workdir, btw). All commands / operations run through `sh` are relative to that working directory. `pushdir` changes that directory, and returns a guard. When that guard is dropped, the directory is popped.

You may be wary if you have bitten by code like this before:

```rust
// guard is immediately dropped, because thats what `let _ = ` does! Oops!
let _ = thing_that_returns_guard();
```

In the next version of xshell, it looks like `pushdir` is going away entirely, and will be replaced with approximately this:

```rust
sh.with_dir(some_dir, || {

})
```

I'll deal with that when it happens. I'm sure the lambdas will make some things more annoying, but I think it'll be better that way.

faithanalog commented:
@davepacheco tagged for review since we've been talking about this for awhile. would also tag ry, but he is out. not sure who else might be interested in reviewing.

davepacheco commented:
Sorry for the delay in looking at this!  I'm just starting to look now.

> Additionally, because of cloning the source to a temporary working path, any changes you have not git added will not be seen by a4x2-package. I consider this either mildly useful or mildly confusing depending on what you're doing in the moment. The jj crew can of course just run jj before using a4x2-package.
...
> Maybe we should just try to make a4x2 work in the current workdir without causing trouble.

I think you've made the right call here in using Git.  I also think making a4x2 work in the current work tree is a great goal, but shouldn't block this work.  In the meantime, I'd suggest that the tool fail by default with a clear message if the working tree is dirty, but allow that to be overridden.

davepacheco commented on dev-tools/xtask/Cargo.toml:
```diff
@@ -31,8 +31,13 @@ clap.workspace = true
 fs-err.workspace = true
 macaddr.workspace = true
 serde.workspace = true
+serde_json.workspace = true
```
There was interest in keeping `xtask` as tiny as possible and using the "external" pattern (which is also pretty easy) to have it invoke other commands instead.

I don't have an opinion about whether this should be "external" or not.

davepacheco commented on dev-tools/xtask/src/a4x2_package.rs:
```diff
@@ -0,0 +1,382 @@
+
+#[derive(Parser)]
+pub struct A4x2PackageArgs {
+    /// Bundle omicron live tests into the output tarball
+    #[clap(long)]
+    live_tests: bool,
+
+    /// Bundle omicron end-to-end-tests package into the output tarball
+    #[clap(long)]
+    end_to_end_tests: bool,
```
Do you have a sense of how much more time these add?  I'd be inclined to make them on by default.

davepacheco commented on dev-tools/xtask/src/a4x2_package.rs:
```diff
@@ -0,0 +1,382 @@
+
+#[derive(Deserialize, Debug)]
+struct NextestConfigVersion {
+    // required: String,
```
Remove?

davepacheco commented on dev-tools/xtask/src/a4x2_package.rs:
```diff
@@ -0,0 +1,382 @@
+    recommended: String,
+}
+
+struct Environment {
+    cargo: String,
+    home_dir: Utf8PathBuf,
+    work_dir: Utf8PathBuf,
+    src_dir: Utf8PathBuf,
+    out_dir: Utf8PathBuf,
+    omicron_dir: Utf8PathBuf,
+    in_ci: bool,
+}
```
Some doc comments here would be a big help.

davepacheco commented on dev-tools/xtask/src/a4x2_deploy.rs:
```diff
@@ -0,0 +1,472 @@
+    // TODO: do we want to have a configurable falcon dataset?
+
+    let env = {
+        // Buildomat sets the CI environment variable. We'll work a bit
+        // differently in that case (no cleanup, use /work)
+        let in_ci = env::var("CI").is_ok();
```
Flags like this make me worried about diverging behavior between CI and local operation.  How about allowing the work directory to be specified on the command line and having CI specify it?  I'm not that worried about skipping cleanup in CI but if we that we could have a `--skip-cleanup` arg, too.

sunshowers replied:
Having read through the rest of this PR, I'm coming around to believing that maybe configuration profiles wouldn't be the worst after all.

davepacheco commented on dev-tools/xtask/src/a4x2_package.rs:
```diff
@@ -0,0 +1,382 @@
+        let work_dir = if in_ci {
+            Utf8PathBuf::from("/work/a4x2-package")
+        } else {
+            home_dir.join(".cache/a4x2-package")
```
Why put this under `$HOME` instead of `target` or `out` or something?

This reminds me: I'd like to have some documentation for this pair of tools, maybe even in the top level README.adoc or a new doc similar to the `docs/how-to-run*.adoc` docs.  It should be clear about what paths it uses and for what.

davepacheco commented on dev-tools/xtask/src/a4x2_package.rs:
```diff
@@ -0,0 +1,382 @@
+        let src_dir = Utf8PathBuf::try_from(env::current_dir()?)?;
+
+        // Delete any results from previous runs. We don't mind if there's errors.
+        fs::remove_dir_all(&work_dir).ok();
```
Worth a confirmation here?  Or: would it make sense to create a new directory in `$work_dir` for this invocation so that we don't have to worry about colliding with a previous one?

davepacheco commented on dev-tools/xtask/src/a4x2_package.rs:
```diff
@@ -0,0 +1,382 @@
+    // Get a ref we can checkout from the local working tree
+    let treeish = {
+        let _popdir = sh.push_dir(&env.src_dir);
+        let mut treeish = cmd!(sh, "git stash create").read()?;
```
This is where I'd be inclined to:

- check if the working tree is dirty
- if so: bail unless some `--allow-dirty` is specified
- regardless, use `HEAD`

sunshowers replied:
Also:

* Use `$GIT` before `git`.
* Altering stash-related state is a visible operation and I don't think we should do it -- instead, there are ways to create a "background" commit from the current state.
* What happens if this is run in a non-git repo, e.g. a non-colocated Jujutsu repo?

Another option that has somewhat less source control interaction is:

* get the list of all tracked files in the repo ([here's instructions on how](https://github.com/camino-rs/paths0/wiki/Source-control-command-list:-git,-hg,-jj))
* create hardlinks for all of them in a mirrored structure. (Hardlinks are fine because basically all editors write files atomically, which breaks them.)

iliana replied:
> Altering stash-related state is a visible operation and I don't think we should do it

`git stash create` (not to be confused with `git stash push`) doesn't put the created commit object anywhere in the ref namespace, per the man page -- is it visible in some other way that creating a tree with `git update-index` etc isn't?

davepacheco commented on dev-tools/xtask/src/a4x2_package.rs:
```diff
@@ -0,0 +1,382 @@
+    cmd!(sh, "git fetch origin {treeish}").run()?;
+    cmd!(sh, "git checkout {treeish}").run()?;
+
+    // XXX is this needed in CI?
```
```suggestion
```

davepacheco commented on dev-tools/xtask/src/a4x2_package.rs:
```diff
@@ -0,0 +1,382 @@
+}
+
+fn build_end_to_end_tests(sh: &Shell, env: &Environment) -> Result<()> {
+    // I'm not confident this does what it should, because I don't know much
+    // about the end to end tests. This is just going from what the shell script
+    // did.
```
@iliana maybe could answer this?

davepacheco commented on dev-tools/xtask/src/a4x2_package.rs:
```diff
@@ -0,0 +1,382 @@
+    // We need nextest available, both to generate the test bundle and to
+    // include in the output tarball to execute the bundle on a4x2.
+
+    // XXX right now we only download a new nextest in CI, and we use
+    // the user's locally installed nextest otherwise. I'm inclined to
+    // say we should always download a fresh one for consistency.
+    // thoughts?
```
```suggestion
    // XXX right now we only download a new nextest in CI, and we use
    // the user's locally installed nextest otherwise. I'm inclined to
    // say we should always download a fresh one for consistency.
    // thoughts?
```

I'd be inclined to have this tool assume that the user has already set up nextest however they want.  That's what the rest of Omicron currently does, I believe.  If CI needs to download it, that's fine, but it should do it as a separate step before invoking this tool (which I think is what it does for the existing CI jobs).

sunshowers replied:
Agreed -- CI already downloads nextest. I'd just package up whatever nextest is available in the `PATH`.

davepacheco commented on dev-tools/xtask/src/a4x2_package.rs:
```diff
@@ -0,0 +1,382 @@
+    )
+    .run()?;
+
+    // XXX use main branch
+    // cmd!(sh, "git -C {testbed_dir} checkout artemis/init-in-smf").run()?;
```
Could make this a command-line argument.  It could accept a branch name or alternatively a local path to a clone if you want to do anything other than "main".

davepacheco reviewed (commented):
This is looking great!  The description in the PR was very helpful in getting oriented -- thanks for that.

I haven't yet looked at `a4x2-deploy`.

davepacheco commented on dev-tools/xtask/src/a4x2_deploy.rs:
```diff
@@ -0,0 +1,472 @@
+pub fn run_cmd(args: A4x2DeployArgs) -> Result<()> {
+    let sh = Shell::new()?;
+
+    // TODO: do we want to have a configurable falcon dataset?
```
Yes, definitely.  I can't run a4x2 on my system without overriding the default here.

It'd be nice to override a few other things, like THREE_NODE_RSS.  Not a blocker though for v1.

davepacheco commented on dev-tools/xtask/src/a4x2_deploy.rs:
```diff
@@ -0,0 +1,472 @@
+                // Delete any results from previous runs. We don't mind if
+                // there's errors. This needs to run as root in case there are
+                // artifacts owned by root left around from the deploy.
+                cmd!(sh, "pfexec rm -rf").arg(&env.work_dir).run()?;
```
This is (obviously) incredibly dangerous.  I wonder how we can make it less so.

sunshowers replied:
Yeah, agreed -- if you could set `work_dir` to something like `/etc` (e.g. via the CLI option discussed above) you could nuke the whole system.

How many files is this typically, and how many are owned by root? If it's not too many, I'm wondering if we could do a first pass that iterates over the file system to find files owned by root, and then lists out those files. Then the user can be prompted to ask if it's okay to use `pfexec` (and there can be a `--force` option to skip prompting.)

davepacheco commented on dev-tools/xtask/src/a4x2_deploy.rs:
```diff
@@ -0,0 +1,472 @@
+    exec_remote_script(&sh, "https://raw.githubusercontent.com/oxidecomputer/falcon/main/get-ovmf.sh")?;
+    exec_remote_script(&sh, "https://raw.githubusercontent.com/oxidecomputer/falcon/main/get-propolis.sh")?;
+
+    // Generate an ssh key we will use to log into the sleds.
+    cmd!(sh, "ssh-keygen -t ed25519 -N '' -f a4x2-ssh-key").run()?;
```
Ooh, (I think) I like using a one-off key here.  But it also needs to be easy for folks to ssh in after the fact.  Should we give folks a command to load it into their agent?

sunshowers replied:
Or provide an ssh command to copy-paste.

davepacheco commented on dev-tools/xtask/src/a4x2_deploy.rs:
```diff
@@ -0,0 +1,472 @@
+        .find(|v| v["dynamic"] == Value::Bool(true))
+        .ok_or(anyhow!("failed to find customer edge addr"))?["local"]
+        .as_str()
+        .unwrap();
```
May as well replace these `unwrap()`s with error messages?
Would it make sense to define a Rust struct that derives `Deserialize` so that if any of the assumptions in `ce_addr_json[0]["addr_info"].as_array().unwrap()` aren't true, we fail gracefully?  (value might not be an array; might not have `addr_info` property, it might not be an array, etc.)

davepacheco commented on dev-tools/xtask/src/a4x2_deploy.rs:
```diff
@@ -0,0 +1,472 @@
+        .as_str()
+        .unwrap();
+
+    // XXX Im told that pinging the gateway from inside the sleds is no longer
+    // necessary so I'm leaving it out for now. We'll see if that's true.
+    // So far it seems to be.
+
```
```suggestion
```

davepacheco commented on dev-tools/xtask/src/a4x2_deploy.rs:
```diff
@@ -0,0 +1,472 @@
+    // necessary so I'm leaving it out for now. We'll see if that's true.
+    // So far it seems to be.
+
+    cmd!(sh, "pfexec route add 198.51.100.0/24 {customer_edge_addr}").run()?;
```
I'm surprised you have to do this because I don't remember doing it.  Where does this happen during a manual a4x2 launch?

sunshowers replied:
(and more generally, could there be a reference to documentation for this function so we can look at things step-by-step?)

davepacheco commented on dev-tools/xtask/src/a4x2_deploy.rs:
```diff
@@ -0,0 +1,472 @@
+    // XXX this is an arbitrary timeout. Should it be configurable? Skippable?
+    let mut retries = 50;
+
+    // The important thing here is to do an HTTP request with timeout to the
+    // control plane API endpoint. If the server is up, we'll get the page you'd
+    // expect for an unauthenticated user. If not, then the request will either
+    // - fail
+    // - stall
+    //
+    // Stalling is why we do a timeout, and we need to handle timeout both at
+    // the TCP level and the HTTP level. `curl -m 5` will hard-out after 5
+    // seconds no matter what's going on.
+    //
+    // That 5 seconds is arbitrary, but it's been working well over in
+    // rackletteadm, from which this logic is copied.
+    //
+    // We could replace curl with rust code, but make sure we do the same thing!
+    while retries > 0 && cmd!(sh, "curl -s -m 5 {api_url}").run().is_err() {
+        retries -= 1;
+        thread::sleep(time::Duration::from_secs(25));
+
+        if retries % 5 == 0 {
+            // I see no reason to error while printing the date
+            let _ = cmd!(sh, "date").run();
+        }
+    }
```
I wonder if you can use the same code that the end-to-end tests use.  I seem to recall them doing the same thing in Rust.

sunshowers replied:
If we do that, then it's worth factoring that out so we don't have to build e.g. our Diesel code to build this binary. (And if we do that, we should definitely put this into an external binary.)

sunshowers commented on dev-tools/xtask/src/a4x2_deploy.rs:
```diff
@@ -0,0 +1,472 @@
+        let home_dir = Utf8PathBuf::from(env::var("HOME")?);
+        let work_dir = if in_ci {
+            Utf8PathBuf::from("/work/a4x2-deploy")
+        } else {
+            home_dir.join(".cache/a4x2-deploy")
+        };
```
I'm a bit worried about files owned by root inside the user-level cache directory, though maybe this is already the case today and we aren't making anything worse.

Trying to think of alternatives:

* there's the Cargo target directory, though files owned by root in there would break `cargo clean`
* somewhere in `/tmp`, though temporary file cleaners might interfere (not sure if illumos has one)

If it must be here, it's worth adding a comment about that.

sunshowers commented on dev-tools/xtask/src/main.rs:
```diff
@@ -40,6 +42,12 @@ struct Args {
 
 #[derive(Subcommand)]
 enum Cmds {
+    /// Generate a tarball with omicron packaged for deployment onto a4x2
+    A4x2Package(a4x2_package::A4x2PackageArgs),
+
+    /// Run a4x2 and deploy omicron onto it, and optionally run live-tests and end to end tests
+    A4x2Deploy(a4x2_deploy::A4x2DeployArgs),
+
```
It's worth gating these on illumos and printing a nice error on other platforms.

sunshowers commented on dev-tools/xtask/src/a4x2_deploy.rs:
```diff
@@ -0,0 +1,472 @@
+            // that in the case of running tests, we will produce an error if
+            // launching a4x2 succeeds but the tests themselves fail.
+            let result =
+                try_launch_a4x2(&sh, &env).and_then(|_| match &args.command {
+                    DeployCommand::RunTests(test_args) => {
+                        run_tests(&sh, &env, &test_args)
+                    }
+                    _ => Ok(()),
+                });
```
There are two sources of errors here: launching a4x2, and running tests. Could you add contexts for each error?

sunshowers commented on dev-tools/xtask/src/a4x2_deploy.rs:
```diff
@@ -0,0 +1,472 @@
+                    }) => {
+                        // We ignore any error here, because the result we
+                        // actually want to produce is whether tests passed
+                        let _ = teardown_a4x2(&sh, &env);
```
Should print a warning here if this failed.

sunshowers commented on dev-tools/xtask/src/a4x2_deploy.rs:
```diff
@@ -0,0 +1,472 @@
+            // (and perhaps others?), to help diagnose a4x2 launch failures, and
+            // provide context alongside live tests.
+
+            // If we are in CI, we do not need to tear down, because the system
+            // state is about to be wiped from existence anyway.
+            if !env.in_ci {
```
Could CI just pass in `--leave-running` instead?

sunshowers commented on dev-tools/xtask/src/a4x2_deploy.rs:
```diff
@@ -0,0 +1,472 @@
+    cmd!(sh, "banner 'prepare'").run()?;
+
+    // TODO could move these into the a4x2-package bundle.
+    // mostly of matter of having a consistent version of propolis.
```
```suggestion
    // mostly a matter of having a consistent version of propolis.
```

sunshowers commented on dev-tools/xtask/src/a4x2_deploy.rs:
```diff
@@ -0,0 +1,472 @@
+
+    // Not sure how this IP is fixed, but it is
+    let api_url = "http://198.51.100.23";
+    println!("polling control plane for signs of life for up to 25 minutes");
```
25 minutes 😅 How long does this usually take? Something like 2x p90 might make sense.

sunshowers commented on dev-tools/xtask/src/a4x2_deploy.rs:
```diff
@@ -0,0 +1,472 @@
+
+    // Timeout = (retries / 2) minutes
+    // XXX this is an arbitrary timeout. Should it be configurable? Skippable?
+    let mut retries = 50;
```
Thinking about it, I feel like if the control plane doesn't come up we shouldn't tear down a4x2 (at least by default) -- instead, we should tell you that we aren't doing so, and then also tell you to run the stop command.

sunshowers commented on dev-tools/xtask/src/a4x2_deploy.rs:
```diff
@@ -0,0 +1,472 @@
+        thread::sleep(time::Duration::from_secs(25));
+
+        if retries % 5 == 0 {
+            // I see no reason to error while printing the date
```
```suggestion
            // No reason to error while printing the date
```

sunshowers commented on dev-tools/xtask/src/a4x2_deploy.rs:
```diff
@@ -0,0 +1,472 @@
+    // rackletteadm, from which this logic is copied.
+    //
+    // We could replace curl with rust code, but make sure we do the same thing!
+    while retries > 0 && cmd!(sh, "curl -s -m 5 {api_url}").run().is_err() {
```
```suggestion
    while retries > 0 && cmd!(sh, "curl --silent --max-time 5 {api_url}").run().is_err() {
```

In general, I think scripts should use long options.

sunshowers commented on dev-tools/xtask/src/a4x2_deploy.rs:
```diff
@@ -0,0 +1,472 @@
+
+        if retries % 5 == 0 {
+            // I see no reason to error while printing the date
+            let _ = cmd!(sh, "date").run();
```
Should also print how much longer we'll be waiting.

sunshowers commented on dev-tools/xtask/src/a4x2_deploy.rs:
```diff
@@ -0,0 +1,472 @@
+
+        let mut had_gateway = false;
+        for ln in route.lines() {
+            if ln.contains("gateway") {
```
Could you add a comment here with an example of what one of these lines looks like?

sunshowers commented on dev-tools/xtask/src/a4x2_deploy.rs:
```diff
@@ -0,0 +1,472 @@
+        let mut had_gateway = false;
+        for ln in route.lines() {
+            if ln.contains("gateway") {
+                let gateway = ln.split_whitespace().nth(1).ok_or(anyhow!(
+                    "teardown_a4x2: could not get gateway for a4x2 route from line {ln}"
+                ))?;
```
I think this can use the anyhow `.with_context` on `Option<T>`.

Also, could you print out the line number via the `enumerate` combinator? And put the line itself in quotes or backticks to clearly demarcate it.

sunshowers commented on dev-tools/xtask/src/a4x2_deploy.rs:
```diff
@@ -0,0 +1,472 @@
+                    "teardown_a4x2: could not get gateway for a4x2 route from line {ln}"
+                ))?;
+
+                cmd!(sh, "pfexec route delete 198.51.100.0/24 {gateway}")
+                    .run()?;
```
Worth printing out the fact that we're doing this.

In general I really like the approach where we first gather all operations into a structure, and then perform them in a separate pass. That enables prompting for confirmation, and things like `--dry-run`. Not sure if you want to put the effort into doing that, but if you do I'd really appreciate it :)

sunshowers commented on dev-tools/xtask/src/a4x2_deploy.rs:
```diff
@@ -0,0 +1,472 @@
+            }
+        }
+
+        if !had_gateway {
+            break;
+        }
```
Is `had_gateway` not being present an exceptional condition? Worth adding a comment in any case.

sunshowers commented on dev-tools/xtask/src/a4x2_deploy.rs:
```diff
@@ -0,0 +1,472 @@
+
+    let ssh_key_path = env.work_dir.join("a4x2-ssh-key");
+
+    // pour one out for enjoyers of clean indentation
+    println!(
```
https://docs.rs/indoc if you'd like to use it!

sunshowers commented on dev-tools/xtask/src/a4x2_deploy.rs:
```diff
@@ -0,0 +1,472 @@
+                    "teardown_a4x2: could not get gateway for a4x2 route from line {ln}"
+                ))?;
+
+                cmd!(sh, "pfexec route delete 198.51.100.0/24 {gateway}")
```
It looks like some of these IPs and subnets are fixed and well-known -- could you make them constants at the top of the file?

sunshowers commented on dev-tools/xtask/src/a4x2_deploy.rs:
```diff
@@ -0,0 +1,472 @@
+
+    for ln in ipadm.lines() {
+        if ln.contains("dhcp") {
+            let ipv4 = ln.split_whitespace().nth(3).ok_or(anyhow!("get_host_ip: could not extract IP for node {node} from line {ln}"))?;
+            let ipv4 = ipv4.strip_suffix("/24").ok_or(anyhow!("get_host_ip: could not extract IP for node {node} from line {ln}"))?;
```
same comments as in the gateway section above.

sunshowers reviewed (commented):
Thank you for doing this! Really excited about automating all of this.

(I'm still working on the review -- here's what I have for now.)

jclulow commented on dev-tools/xtask/src/a4x2_deploy.rs:
```diff
@@ -0,0 +1,472 @@
+    for _ in 0..10 {
+        let mut route_cmd = cmd!(sh, "pfexec route get 198.51.100.0/24");
+
+        // We get an error code when there is no route (and thus, nothing for
+        // us to delete!)
```
I think before ignoring the result here I would make sure the output ends in **not in table**, otherwise it's presumably an error we're _not_ expecting and we'd like to display it to the user and abort.

jclulow commented on dev-tools/xtask/src/a4x2_deploy.rs:
```diff
@@ -0,0 +1,472 @@
+                        leave_running: false,
+                        ..
+                    }) => {
+                        // We ignore any error here, because the result we
+                        // actually want to produce is whether tests passed
+                        let _ = teardown_a4x2(&sh, &env);
+                    }
```
I don't think we should ignore this error.  Either **teardown_a4x2()** should work, or we need to stop and figure out why it did not.

jclulow commented on dev-tools/xtask/src/a4x2_deploy.rs:
```diff
@@ -0,0 +1,472 @@
+        DeployCommand::Stop => teardown_a4x2(&sh, &env)?,
+        DeployCommand::Start(_) | DeployCommand::RunTests(_) => {
+            if !env.in_ci {
+                // Teardown previous deploy if it exists, before wiping the data
+                // for it. If this errors, we assume the deploy doesn't exist
+                // and carry on.
+                let result = teardown_a4x2(&sh, &env);
```
I don't think we should ignore this error.  Either **teardown_a4x2()** should work, or we need to stop and figure out why it did not.

jclulow reviewed (commented):
This seems like a great idea!  I have a hodge podge of thoughts I've left as comments, but I wanted to say I agree this makes things much more accessible to casual users in particular.  I also like the concrete separation of build machine and target machine, with an artefact you can pass between them!

sunshowers commented on dev-tools/xtask/src/a4x2_package.rs:
```diff
@@ -0,0 +1,382 @@
+    // I'm not confident this does what it should, because I don't know much
+    // about the end to end tests. This is just going from what the shell script
+    // did.
+    cmd!(sh, "banner 'end to end'").run()?;
```
hmm, I know `banner` has a character limit, might be worth checking whether this command works.

faithanalog replied:
it does, i ran into that char limit earlier in dev but this one is fine

sunshowers commented on dev-tools/xtask/src/a4x2_package.rs:
```diff
@@ -0,0 +1,382 @@
+            }
+        }
+
+        if can_dedupe {
+            // yay, extract out to the common dir.
+            let dest = common_dir.join(base_path);
+            sh.create_dir(&dest.parent().unwrap())?;
+            sh.copy_file(&g0_path, &dest)?;
+            sh.remove_path(&g0_path)?;
+            for prefix in &prefixes {
+                sh.remove_path(&prefix.join(base_path))?;
+            }
+        }
```
Hmm interesting -- what code ensures that sled-common is checked?

faithanalog replied:
In a4x2-deploy, I've got this block of commands

```
    cmd!(sh, "tar -cf sled-common.tar -C {a4x2_dir}/cargo-bay/sled-common ./")
        .run()?;
    cmd!(sh, "tar -xf sled-common.tar -C {a4x2_dir}/cargo-bay/g0").run()?;
    cmd!(sh, "tar -xf sled-common.tar -C {a4x2_dir}/cargo-bay/g1").run()?;
    cmd!(sh, "tar -xf sled-common.tar -C {a4x2_dir}/cargo-bay/g2").run()?;
    cmd!(sh, "tar -xf sled-common.tar -C {a4x2_dir}/cargo-bay/g3").run()?;
```

This
- creates an intermediate tar of the common dir
- extracts it into each of g0/g1/g2/g3's individual directories

which, since tar does not delete any of the files in `g{0,1,2,3}`, this simply adds them in place in addition to anything else there

sunshowers commented on dev-tools/xtask/src/a4x2_package.rs:
```diff
@@ -0,0 +1,382 @@
+            .run()?;
+        env.home_dir.join(".cargo/bin/cargo-nextest")
+    } else {
+        // XXX do we need to test nextest exists & is the right version? need to
+        // see if the incidental errors from stuff below already take care of
+        // telling someone outside of CI what they need to be doing.
```
You could run `cargo nextest show-config version` -- [exit code 92](https://docs.rs/nextest-metadata/latest/nextest_metadata/enum.NextestExitCode.html#associatedconstant.REQUIRED_VERSION_NOT_MET) means that the required version isn't met, and [exit code 10](https://docs.rs/nextest-metadata/latest/nextest_metadata/enum.NextestExitCode.html#associatedconstant.RECOMMENDED_VERSION_NOT_MET) means that the recommended version isn't met.

sunshowers commented on dev-tools/xtask/src/a4x2_package.rs:
```diff
@@ -0,0 +1,382 @@
+    // intentionally relative path so the tarball gets relative paths
+    let _popdir = sh.push_dir(&env.work_dir);
+    let bundle_dir_name = live_test_bundle_dir.file_name().unwrap();
+    cmd!(sh, "tar -czf {out_dir}/live-tests-bundle.tgz {bundle_dir_name}/")
```
`live-tests-bundle.tgz` maybe can be a constant.

sunshowers commented on dev-tools/xtask/src/a4x2_package.rs:
```diff
@@ -0,0 +1,382 @@
+    let pkg_dir_name = env.out_dir.file_name().unwrap();
+
+    // Place output in `out/` when running locally, or leave in work dir for CI
+    let artifact = Utf8PathBuf::from("a4x2-package-out.tgz");
```
This filename can be a constant too, I think.

sunshowers commented on dev-tools/xtask/src/a4x2_package.rs:
```diff
@@ -0,0 +1,382 @@
+    let a4x2_dir = testbed_dir.join("a4x2");
+
+    // TODO move this clone & build to prepare, so that if the omicron build
+    // takes awhile we wont have lost our git token when we get here. Relevant
```
```suggestion
    // takes awhile, we won't have lost our git token when we get here. Relevant
```

faithanalog commented on dev-tools/xtask/src/a4x2_deploy.rs:
```diff
@@ -0,0 +1,472 @@
+    let api_url = "http://198.51.100.23";
+    println!("polling control plane for signs of life for up to 25 minutes");
+
+    // Print the date for the logs' benefit
+    let _ = cmd!(sh, "date").run();
+
```
sadly, no

faithanalog commented on dev-tools/xtask/src/a4x2_deploy.rs:
```diff
@@ -0,0 +1,472 @@
+
+    cmd!(sh, "pfexec route add 198.51.100.0/24 {customer_edge_addr}").run()?;
+
+    // Not sure how this IP is fixed, but it is
+    let api_url = "http://198.51.100.23";
```
probably the case, though I am not sure where this IP range comes from so I am not sure how to parameterize it yet. it may be from some file in the testbed repo

faithanalog commented on dev-tools/xtask/src/a4x2_deploy.rs:
```diff
@@ -0,0 +1,472 @@
+
+        cmd!(sh, "scp {ssh_args...} live-tests-bundle.tgz {ssh_host}:/zone/oxz_switch/root/root").run()?;
+
+        // If you want any change in functionality for the test runner, update
+        // run-live-tests over in a4x2_package.rs. Don't add it here!
+        let switch_zone_script = r#"
+            set -euxo pipefail
+            tar xvzf live-tests-bundle.tgz
+            cd live-tests-bundle
+            ./run-live-tests
+        "#;
+
+        let remote_script =
+            format!("zlogin oxz_switch bash -c '{switch_zone_script}'");
+
+        // Will error if the live tests fail. This is desired.
+        cmd!(sh, "ssh {ssh_args...} {ssh_host} {remote_script}").run()?;
```
xshell doesn't invoke shell, it's just rust Command, so no escaping necessary there. ssh will feed the zlogin command as is to the shell so that's also fine.

there is one place that this is at risk of escape behavior, which is if the script contains any single-quotes in it. I can fix that with ye olde `replace("'", "'\"'\"'")` and then it will be clean all the way thru even if someone starts changing the script in arbitrary ways.

but I agree that I am not thrilled by what's going on here.

jgallagher commented:
> When a4x2 succeeds, control plane fails 11.68% of the time. meaning: a4x2 turned on fine, but we timed out waiting for the API to come up.

FWIW (maybe very little!): #7712 fixes a problem that has caused this failure mode for me in a4x2 with some regularity. This was a real bug that could occur on real systems, but is much less likely on a healthy system. I wonder how much of this 11.68% is either that exact issue or something similarly possible-albeit-unlikely on real hardware.

faithanalog commented:
I am declaring github UX bankruptcy for replying to all the comment threads one by one, but to the best of my knowledge I've addressed everything. But, it has been a lot of changes, so I will not be surprised if folks raise new issues.

## Primary changes since initial PR:

### work is stored in `target/`

No more stuffing things in `~/.cache`. This also makes it `cargo clean`-able. Nothing is owned by root, so I'm not doing any `pfexec rm` anymore.

As root-owned files go, the only thing that was owned by root was the files created by falcon when running `pfexec a4x2 launch`. Right now I'm solving that problem by `chown`-ing them to the current user after `launch` runs. It might be nicer if falcon could be run without root and elevate itself as needed; then it could create those files in advance as a non-root user. I'm putting that under potential future improvements.

### copying omicron source code

I switched to using rsync instead of git. This matches up with how I actually do things in my personal build scripts, and have done for some time.

The actual requirements were:

- clone the source tree, preserving file permissions, symlinks, etc.
- don't copy `out/` or `target/`

It's a minor downside that `rsync` is not installed by default on a helios box, but I think it is common enough to be ok to use here. I also considered doing a `tar | tar`, but I didn't trust myself to use illumos tar correctly.

### no more `in_ci`

everything has either been made to work one way only, or turned into a command line flag.

### we can collect logs from a4x2 now

somehow I forgot to actually implement this logic the first go around... it's there now as `xtask a4x2 deploy collect-evidence`. I collect falcon logs and all the logs I thought to grab from inside the sleds.

### `deploy` usage

`deploy` is structured more like a developer tool that CI will run, rather than a CI tool that developers will run. This primarily means, one command per action:

```
Usage: cargo xtask a4x2 deploy <COMMAND>

Commands:
  install-propolis  Download and install the propolis hypervisor and bootcode. Pre-requisite for launching a4x2
  start             Start a4x2, deploy a control plane to it, and then leave it running. Destroys any existing a4x2 deployment, unless --ensure is used
  stop              Stop a4x2 that was previously launched with the `start` subcommand
  run-live-tests    Execute live-tests previously built by `xtask a4x2 package`
  collect-evidence  Collect logs from a4x2, and services within the sleds
  status            Query a4x2 node access information
```

This allows some flexibility for how to run things if using this tool while iterating. The usual flow would be

```
cargo xtask a4x2 deploy start \
    && cargo xtask a4x2 deploy run-live-tests

cargo xtask a4x2 deploy collect-evidence

cargo xtask a4x2 deploy stop
```

And in fact this is what CI will do. But if you instead wanted to run a4x2 and do something other than run bundled live tests, you can do that too. Previously I was making this possible with command line arguments to `start`, but this seemed less awkward to me.

This also means that `deploy start` polling could be made optional if we wanted. I left it as unconditional because I always want the command to poll for the control plane and tell me if it came up or not, and nobody has complained about the similar behavior in rackletteadm.

### Other changes

- a4x2 commands moved to submodule to lift common code/constants out to a mod.rs
- a4x2 is a subcommand
    - `a4x2-deploy` -> `a4x2 deploy`
    - `a4x2-package` -> `a4x2 package`
- lots of additional comments to address things people asked questions about in review
- lots of `context()`
- `fs_err`
- allow specifying testbed source/branch
- add user's ssh pubkeys to sleds
- reduce control plane polling window to 20 minutes. Hopefully this isn't a problem for slower systems?
- print remaining time while polling
- env scrubbing
- `route -n`
- longopts for curl
- `Deserialize` instead of adhoc json parsing
- always build live-tests / end-to-end tests

## Known issues / annoyances

### install-propolis temporarily busted

`xtask deploy install-propolis` will install OVMF, but I have forbidden it from installing propolis for now. See https://github.com/oxidecomputer/falcon/issues/106 for details; this is something I want to fix before merging.


### i am always compiling omicron at all hours

While packaging, we will rebuild `omicron-package` and some parts of omicron multiple times. This is a new issue I picked up during a rebase! It's because of ring, like https://github.com/oxidecomputer/omicron/pull/7740. I don't think this is on me; I'm filtering out the offending env vars on my executions of things. It might be that `omicron-package` needs to filter the variables out, I am not sure. It does not break things, just makes things much slower than they ought to be for the time being.


### `git clone https://` my nemesis

Because the default `testbed` source path is https, folks who usually clone by git (such as myself) may get prompted for username/password. It's the first thing that happens when running `a4x2 package`, so it's at least not a nasty surprise to come back to after leaving the computer for a bit. I've just been specifying the git version of the source path while testing, but I want to figure out how to not need to do that.


### if propolis core-dumps that's annoying

If propolis-server core dumps it does it in the work directory, and then you have a file owned by root you need to manually delete. I'm tempted to call this one out of scope, but I could look for a `core` file and add it to the list of things I chown if folks want.

davepacheco commented on dev-tools/xtask/src/a4x2/a4x2_package.rs:
```diff
@@ -0,0 +1,399 @@
+
+/// This script will be placed in the live tests bundle to run the live tests
+const LIVE_TESTS_EXECUTION_SCRIPT: &str = r#"
+    set -euxo pipefail
```
nit: I like to use explicit names here so readers don't have to have these all memorized (e.g., `set -o errexit`).

davepacheco commented on dev-tools/xtask/src/a4x2/a4x2_package.rs:
```diff
@@ -0,0 +1,399 @@
+
+        // Delete any results from previous runs.
+        if work_dir.try_exists()? {
+            fs::remove_dir_all(&work_dir)?;
```
suggestion: print a message saying it's doing this

davepacheco commented on dev-tools/xtask/src/a4x2/a4x2_package.rs:
```diff
@@ -0,0 +1,399 @@
+            fs::remove_dir_all(&work_dir)?;
+        }
+
+        // Create work dir
```
same -- print a message saying it's doing this ("using working directory {work_dir}")

davepacheco commented on dev-tools/xtask/src/a4x2/a4x2_package.rs:
```diff
@@ -0,0 +1,399 @@
+        fs::create_dir_all(&work_dir)?;
+        sh.change_dir(&work_dir);
+
+        // Output. Maybe in CI we want this to be /out
```
```suggestion
        // Output.
```

davepacheco commented on dev-tools/xtask/src/a4x2/a4x2_package.rs:
```diff
@@ -0,0 +1,399 @@
+    let _popdir = sh.push_dir(&env.omicron_dir);
+
+    let cargo = &env.cargo;
+    cmd!(sh, "{cargo} build -p end-to-end-tests --bin commtest --bin dhcp-server --release").run()?;
```
Why `--release`?  I just wonder if leaving that off would improve runtime.

davepacheco commented on dev-tools/xtask/src/a4x2/a4x2_package.rs:
```diff
@@ -0,0 +1,399 @@
+
+    // This will modify some omicron configs, build omicron, and package things
+    // up for the a4x2 cargo-bay
+    cmd!(sh, "./config/build-packages.sh")
+        .env("OMICRON", &env.omicron_dir)
+        .run()?;
```
Will this honor env vars I've set (like FALCON_DATASET)?

davepacheco commented on dev-tools/xtask/src/a4x2/a4x2_deploy.rs:
```diff
@@ -0,0 +1,938 @@
+    /// the command will *ensure* a4x2 is running, starting it if needed, but
+    /// leaving an existing deployment alone.
+    #[clap(long)]
+    ensure: bool,
```
suggestion: call this `--if-not-running`

davepacheco commented on dev-tools/xtask/src/a4x2/a4x2_deploy.rs:
```diff
@@ -0,0 +1,938 @@
+            // for it.
+            if env.falcon_dir.try_exists()? {
+                if args.ensure {
+                    // ensure arg says: make sure a4x2 is in theory deployed.
+                    // right now we will use the presence of .falcon to indicate
+                    // that, but that can be an incorrect if the system rebooted
+                    // since a4x2 was deployed.
+                    eprintln!(".falcon directory exists; assuming a4x2 is up");
+                    return Ok(());
```
It seems like this isn't reliable in that it will only _not_ tear down if the falcon directory is present.  But it might be present, yet nothing is running.  I wonder: what's the motivation for this flag?

davepacheco commented on dev-tools/xtask/src/a4x2/a4x2_deploy.rs:
```diff
@@ -0,0 +1,938 @@
+
+    // Generate an ssh key we will use to log into the sleds.
+    // `-N ''` means "empty passphrase"
+    cmd!(sh, "ssh-keygen -t ed25519 -N '' -f a4x2-ssh-key").run()?;
```
I'd suggest printing a one line summary of what we did here.  Something like `loaded {n} keys from $HOME/.ssh`

davepacheco reviewed (commented):
Very nice!  I also haven't had a chance to look as closely but agree with Joshua's comments that we're close enough at this point and the risk is low enough that I think we should land it sooner rather than later.

The two things I'd maybe consider blockers are:

- this could use some instructions, either in README.adoc in the Omicron root or else a new docs/how-to-run-a4x2.adoc (alongside the existing how-to-run{,-simulated} docs)
- speaking for myself, I need to be able to use FALCON_DATASET -- does that get honored here if I set it in my environment?

jclulow commented:
Oh, there is no actual CI job added with this, right, it's just tools?
